R/read_AHCCD_daily_precip.R

```diff
@@ -0,0 +1,158 @@
+#'adjusted daily precipitation dataset for trend analysis in Canada. 
+#'Atmosphere-Ocean, 49 (2), 163-177.}
+#' @examples
+#' \dontrun{
+#'stoon_rainfall <- read_AHCCD_daily_precip("dr_4057120.txt")}
+#' @export
+
+read_AHCCD_daily_precip <- function(precip_file){
+  # check parameter
+  if (precip_file == "" | is.null(precip_file)){
+    cat('Error: precip_file missing\n')
+    return(FALSE)
+  }
+  
+  # strip off file name from path and extension to figure out data type
+  base <- basename(precip_file) # remove path
+  split <- stringr::str_split_fixed(base, stringr::fixed('.'), 2)
+  filename <- split[1]
+  # check for precip type
+  if (stringr::str_detect(tolower(filename), 'dt')){
+    precip_type <- 'total'
+  }
+  else if (stringr::str_detect(tolower(filename), 'dr')){
+    precip_type <- 'rain'
+  }
+  else if (stringr::str_detect(tolower(filename), 'ds')){
+    precip_type <- 'snow'
+  }
+  else{
+    cat('Error: wrong file type\n')
+    return(FALSE)
+  }
+  
+  # set up homes for data
+  value <- c(0)
+  code <- c(0)
+  date <- c(0)
+  
+  # set up constants
+  twodigitnums <- c('01','02','03','04','05','06','07','08','09','10','11','12',
+                    '13','14','15','16','17','18','19','20','21','22','23','24',
+                    '25','26','27','28','29','30','31')
+  
+  
+  # figure out header info
+  # read 1st 10 lines
+  con <- file(precip_file, "r", blocking = FALSE, encoding="ISO_8859-2")
+  input <- readLines(con, n=10)
+  close(con)
+  
+  # find header lines
+  # first find number of lines containing file info
+  input <- tolower(input)
+  LineNum <- stringr::str_detect(input, stringr::fixed(','))
+  englishHeaderNum <- stringr::str_detect(input, stringr::fixed("updated", ignore_case=TRUE))
+  englishHeaderCount <- sum(englishHeaderNum)
+  frenchHeaderNum <- stringr::str_detect(input, stringr::fixed("jusqu", ignore_case=TRUE))
+  frenchHeaderCount <- sum(frenchHeaderNum)
+  fileHeaderLines <- englishHeaderCount + frenchHeaderCount
+  
+  # now find number of lines containing column titles
+  englishLineNum <- stringr::str_detect(input,stringr::fixed("year", ignore_case=TRUE))
+  englishLineCount <- sum(englishLineNum)
+  frenchLineNum <- stringr::str_detect(input,stringr::fixed("annee", ignore_case=TRUE))
+  frenchLineCount <- sum(frenchLineNum)
+  columnHeaderLines <- sum(englishLineCount) + sum(frenchLineCount)
+  
+  totalSkipLines <- fileHeaderLines + columnHeaderLines
+  
+  # check for with of first field - is there a leading space
+  firstChar <- substr(input[totalSkipLines + 1], 1, 1)
+  
+  if (firstChar == ' ')
+    yearWidth <- 5
+  else
+    yearWidth <- 4
+  
+  # set up widths to read
+  header <- c(yearWidth,3,1)
+  header_classes <- c('numeric','numeric','character')
+  
+  cols <- rep.int(c(8,1),31)
+  cols_classes <- rep.int(c('numeric', 'character'), 31)
+  all <- c(header,cols)
+  all_classes <- c(header_classes, cols_classes)
+  
+  # read data
```

Comment:
One other option here is to replace much of this code with a combination of `tidyr` `gather` and `spread`